<a href="https://colab.research.google.com/github/cardstud/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/blob/master/module1-join-and-reshape-data/LS_DS_121_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-09-19 21:41:29--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 54.231.50.106
Connecting to s3.amazonaws.com (s3.amazonaws.com)|54.231.50.106|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  75.9MB/s    in 2.6s    

2019-09-19 21:41:32 (75.9 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [0]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [0]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [0]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
!curl https://gist.githubusercontent.com/jeremystan/c3b39d947d9b88b3ccff3147dbcf6c6b/raw/0a6a6b96c80016cd1e44d8963745ce66196245e0/data_description.md

`orders` (3.4m rows, 206k users):
* `order_id`: order identifier
* `user_id`: customer identifier
* `eval_set`: which evaluation set this order belongs in (see `SET` described below)
* `order_number`: the order sequence number for this user (1 = first, n = nth)
* `order_dow`: the day of the week the order was placed on
* `order_hour_of_day`: the hour of the day the order was placed on
* `days_since_prior`: days since the last order, capped at 30 (with NAs for `order_number` = 1)

`products` (50k rows):
* `product_id`: product identifier
* `product_name`: name of the product
* `aisle_id`: foreign key
* `department_id`: foreign key

`aisles` (134 rows):
* `aisle_id`: aisle identifier
* `aisle`: the name of the aisle

`deptartments` (21 rows):
* `department_id`: department identifier
* `department`: the name of the department

`order_products__SET` (30m+ rows):
* `order_id`: foreign key
* `product_id`: foreign key
* `add_to_cart_order`: order in which each product was added to cart
* `reo

In [0]:
# Let's use products.csv and use product_id and product_name along with order_products_prior.csv with product_id

In [0]:
import pandas as pd

In [0]:
products = pd.read_csv('products.csv')
products.head()

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [0]:
products.shape

(49688, 4)

In [0]:
products.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id'], dtype='object')

In [0]:
products[products['product_name'] == 'Banana']

product_id product_name  aisle_id  department_id
24851       24852       Banana        24              4

In [0]:
products[products['product_name'] == 'Organic Whole Milk']

product_id        product_name  aisle_id  department_id
27844       27845  Organic Whole Milk        84             16

In [0]:
order_products_prior = pd.read_csv('order_products__prior.csv')
order_products_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [0]:
order_products_prior.columns

Index(['order_id', 'product_id', 'add_to_cart_order', 'reordered'], dtype='object')

In [0]:
order_products_prior.shape

(32434489, 4)

In [0]:
orders = pd.read_csv('orders.csv')
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [0]:
orders.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [0]:
orders.shape

(3421083, 7)

In [0]:
prod = []
lst =  ['Banana', 'Bag of Organic Bananas', 'Organic Strawberries', 'Organic Baby Spinach', 'Organic Hass Avocado', 'Organic Avocado', 'Large Lemon', 'Strawberries', 'Limes', 'Organic Whole Milk']
for x in products():
  if products[products['product_name'] == lst
              prod.append(x)
  

SyntaxError: ignored

In [0]:
top_10 = pd.concat([orders, order_products_prior, products])

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [0]:
top_10.head()

add_to_cart_order  aisle_id  ...  reordered  user_id
0                NaN       NaN  ...        NaN      1.0
1                NaN       NaN  ...        NaN      1.0
2                NaN       NaN  ...        NaN      1.0
3                NaN       NaN  ...        NaN      1.0
4                NaN       NaN  ...        NaN      1.0

[5 rows x 13 columns]

In [0]:
#top_10.groupby('order_id')['product_id'].count().sort_values(ascending=False)
top_10.groupby('order_id')['product_id'].count().sort_values(ascending=False)

order_id
1564244.0    145
790903.0     137
61355.0      127
2970392.0    121
2069920.0    116
3308010.0    115
2753324.0    114
2499774.0    112
77151.0      109
2621625.0    109
1657096.0    108
2136777.0    108
2648316.0    105
171934.0     104
1892299.0    104
1867980.0    102
1384519.0    102
653887.0     102
3282039.0    101
3052353.0    101
2715446.0    100
3048680.0    100
416346.0     100
1101266.0    100
259043.0      99
2716231.0     99
1959075.0     98
3279252.0     98
2944140.0     98
598905.0      98
            ... 
2112500.0      0
2112480.0      0
2112849.0      0
2112467.0      0
2112454.0      0
2112452.0      0
2112441.0      0
344762.0       0
2112413.0      0
2112400.0      0
2112552.0      0
344726.0       0
2112556.0      0
344724.0       0
2112829.0      0
2112806.0      0
2112789.0      0
2112773.0      0
2112754.0      0
2112751.0      0
2112734.0      0
2112681.0      0
2112677.0      0
2112674.0      0
2112657.0      0
2112616.0      0
344713.0       0
21125

In [0]:
#products[products['product_id'] == 24852]
top_10[top_10['order_id'] == 1564244]


add_to_cart_order  aisle_id  ...  reordered  user_id
380638                  NaN       NaN  ...        NaN  22906.0
14825186                1.0       NaN  ...        0.0      NaN
14825187                2.0       NaN  ...        0.0      NaN
14825188                3.0       NaN  ...        1.0      NaN
14825189                4.0       NaN  ...        1.0      NaN
14825190                5.0       NaN  ...        1.0      NaN
14825191                6.0       NaN  ...        1.0      NaN
14825192                7.0       NaN  ...        1.0      NaN
14825193                8.0       NaN  ...        1.0      NaN
14825194                9.0       NaN  ...        1.0      NaN
14825195               10.0       NaN  ...        1.0      NaN
14825196               11.0       NaN  ...        1.0      NaN
14825197               12.0       NaN  ...        1.0      NaN
14825198               13.0       NaN  ...        1.0      NaN
14825199               14.0       NaN  ...        1.0      NaN
14825200               15.0       NaN  ...        1.0      NaN
14825201               16.0       NaN  ...        1.0      NaN
14825202               17.0       NaN  ...        1.0      NaN
14825203               18.0       NaN  ...        1.0      NaN
14825204               19.0       NaN  ...        1.0      NaN
14825205               20.0       NaN  ...        1.0      NaN
14825206               21.0       NaN  ...        1.0      NaN
14825207               22.0       NaN  ...        1.0      NaN
14825208               23.0       NaN  ...        1.0      NaN
14825209               24.0       NaN  ...        1.0      NaN
14825210               25.0       NaN  ...        1.0      NaN
14825211               26.0       NaN  ...        1.0      NaN
14825212               27.0       NaN  ...        1.0      NaN
14825213               28.0       NaN  ...        1.0      NaN
14825214               29.0       NaN  ...        1.0      NaN
...                     ...       ...  ...        ...      ...
14825301              116.0       NaN  ...        0.0      NaN
14825302              117.0       NaN  ...        0.0      NaN
14825303              118.0       NaN  ...        0.0      NaN
14825304              119.0       NaN  ...        0.0      NaN
14825305              120.0       NaN  ...        0.0      NaN
14825306              121.0       NaN  ...        0.0      NaN
14825307              122.0       NaN  ...        0.0      NaN
14825308              123.0       NaN  ...        0.0      NaN
14825309              124.0       NaN  ...        0.0      NaN
14825310              125.0       NaN  ...        0.0      NaN
14825311              126.0       NaN  ...        0.0      NaN
14825312              127.0       NaN  ...        0.0      NaN
14825313              128.0       NaN  ...        0.0      NaN
14825314              129.0       NaN  ...        0.0      NaN
14825315              130.0       NaN  ...        1.0      NaN
14825316              131.0       NaN  ...        0.0      NaN
14825317              132.0       NaN  ...        0.0      NaN
14825318              133.0       NaN  ...        0.0      NaN
14825319              134.0       NaN  ...        0.0      NaN
14825320              135.0       NaN  ...        0.0      NaN
14825321              136.0       NaN  ...        0.0      NaN
14825322              137.0       NaN  ...        0.0      NaN
14825323              138.0       NaN  ...        0.0      NaN
14825324              139.0       NaN  ...        0.0      NaN
14825325              140.0       NaN  ...        0.0      NaN
14825326              141.0       NaN  ...        0.0      NaN
14825327              142.0       NaN  ...        0.0      NaN
14825328              143.0       NaN  ...        0.0      NaN
14825329              144.0       NaN  ...        0.0      NaN
14825330              145.0       NaN  ...        0.0      NaN

[146 rows x 13 columns]

In [0]:
product_top = top_10['product_name'].value_counts()

In [0]:
product_top

Sunscreen Cream, Kids, Tangerine & Vanilla, Broad Spectrum SPF 30                  1
Conditioner Pomegranate + Sunflower                                                1
Dark Melting Wafers for Candy Making & Dipping                                     1
Draft Beer                                                                         1
Drink Distinct All Natural Soda Pineapple Coconut & Nutmeg                         1
Fruit by the Foot, Fruit Roll-Ups, Fruit Gushers Party Pack Fruit Snacks 24 Ct.    1
Frozen Corn On Cob                                                                 1
Melograno e Arancia Sparkling Pomegranate & Orange Beverage                        1
Apis Mellifica 30 C                                                                1
Canned Soda                                                                        1
Original Crackers                                                                  1
Organic Sichuan Peppercorn                                       

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
flights = sns.load_dataset('flights')

In [0]:
##### YOUR CODE HERE #####

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####